agasc/agasc.py

```diff
@@ -397,29 +509,67 @@ def get_stars(ids, agasc_file=None, dates=None, fix_color1=True, use_mag_est=Fal
+for func in get_stars, get_star, get_agasc_cone:
+    func.__doc__ = func.__doc__.format(common_doc=COMMON_DOC)
+
+
+def update_from_supplement(stars):
+    """Update mag, color1 and class information from AGASC supplement in ``stars``.
+
+    Stars with available mag estimates in the AGASC supplement are updated
+    in-place in the input ``stars`` Table:
+
+    - ``MAG_ACA`` and ``MAG_ACA_ERR`` are set according to the supplement.
+    - ``MAG_CATID`` (mag catalog ID) is set to ``MAG_CATID_SUPPLEMENT`` (128).
+    - If COLOR1 is 0.7 or 1.5 then it is changed to 0.69 or 1.49 respectively.
 
-    Stars with available mag estimates in the supplement are updated in-place in
-    the ``stars`` Table. The catalog ID is set to ``MAG_CATID_SUPPLEMENT`` (128).
-    Where COLOR1 is 0.7 or 1.5 it is changed to 0.69 or 1.49 respectively.
+    Stars which are in the bad stars table are updated as follows:
+
+    - ``CLASS = 100 + bad star source id``
+
+    This functionality is gloabally disabled if the environment variable
+    ``AGASC_DISABLE_SUPPLEMENT`` is set to any value.
 
     :param stars: astropy.table.Table of stars
     """
-    for idx, agasc_id in enumerate(stars['AGASC_ID']):
-        agasc_id = int(agasc_id)
-        if agasc_id in SUPPLEMENT:
-            mag_est, mag_est_err = SUPPLEMENT[agasc_id]
-            stars['MAG_ACA'][idx] = mag_est
+    if DISABLE_SUPPLEMENT_ENV in os.environ:
+        return
+
+    def set_star(star, name, value):
+        """Set star[name] = value if ``name`` is a column in the table"""
+        try:
+            star[name] = value
+        except KeyError:
+            pass
+
+    # Get estimate mags and errs from supplement as a dict of dict
+    # agasc_id : {mag_aca: .., mag_aca_err: ..}.
+    supplement_mags = get_supplement_table('mags', as_dict=True)
+
+    # Get bad stars as {agasc_id: {source: ..}}
+    bad_stars = get_supplement_table('bad', as_dict=True)
+
+    for star in stars:
+        agasc_id = int(star['AGASC_ID'])
+        if agasc_id in supplement_mags:
+            mag_est = supplement_mags[agasc_id]['mag_aca']
+            mag_est_err = supplement_mags[agasc_id]['mag_aca_err']
+
+            set_star(star, 'MAG_ACA', mag_est)
             # Mag err is stored as int16 in units of 0.01 mag. Use same convention here.
-            stars['MAG_ACA_ERR'][idx] = round(mag_est_err * 100)
-            stars['MAG_CATID'][idx] = MAG_CATID_SUPPLEMENT
-            color1 = stars['COLOR1'][idx]
-            if np.isclose(color1, 0.7) or np.isclose(color1, 1.5):
-                stars['COLOR1'][idx] = color1 - 0.01
+            set_star(star, 'MAG_ACA_ERR', round(mag_est_err * 100))
+            set_star(star, 'MAG_CATID', MAG_CATID_SUPPLEMENT)
+            if 'COLOR1' in stars.colnames:
+                color1 = star['COLOR1']
+                if np.isclose(color1, 0.7) or np.isclose(color1, 1.5):
+                    star['COLOR1'] = color1 - 0.01
```

Comment:
This seems fine overall, though I note it may be a little confusing for / will not apply to stars that have already been "fixed up" by the update_color1_column method.

Reply:
Good catch on a possible interaction that I didn't notice, but it looks like this is fine. The color update runs first and might change the 1.5 to something else, in which case this code won't do anything, which is fine. Basically we just need to un-1.5-ify that color one way or another. If only the AGASC had come with another status column 25 years ago we wouldn't be playing these games.

Reply:
Right, I mean, we talked a bit about updating starcheck and proseco to just not *use* color1  but I still don't have any great alternative ideas at this point anyway.

Reply:
The real truth is that COLOR1 is already effectively a status column. In other words, we don't actually use the value of color anywhere in operations or analysis apart from checking for 0.7 and 1.5, and of course estimating MAG_ACA in the AGASC catalog generation. From the operations perspective it is basically a 3-valued status flag, "OK", "Unreliable mag - no color", and "Unreliable mag - red". 

We could formalize this by defining exported constants in `agasc`, if that would help things, but the name of this "status" column will always be misleading.

Reply:
Good points.  I was thinking more that for me the confusion is that the 'OK' status can be either 'OK color in original catalog', 'OK color due to update from AGASC 1.7 color information', 'OK color from due to mag estimates previous observations'.  Are all of the MAG_ACA_ERR estimates equally good for those 3 "OK" cases?  

As a tiny nit, I suppose this brings up the issue that I had pushed to get the mag supplement to have an entry for all color1=1.5 stars if observed.  I think if the mag supplement code is identifying color1=1.5 stars based on the agasc get_star method with fix_color1=True, we don't automatically get new mag estimates for stars that were updated via the AGASC 1.7 color updates.  Do we care?  Maybe we don't.  We get new estimates for all stars if "different enough" from catalog and the "color1 updated to be not 1.5 due to AGASC 1.7" stars already had their color1 penalties removed in star selection.

Reply:
> Are all of the MAG_ACA_ERR estimates equally good for those 3 "OK" cases?

Yes. Note that I don't see any reason to ever use `fix_color1=False` in operations or analysis code. It's there for testing / development to see what's really in the AGASC.

> I had pushed to get the mag supplement to have an entry for all color1=1.5 stars if observed

The mag supplement now has an entry for *every* star if it has been observed, and we are not applying any "different enough" criteria for inclusion. That was an early idea but it got dropped IIRC. At least this is all my understanding, given that the supplement has 81000 stars in it and most of them are very close to the original.

Reply:
Ah.  OK.  I guess I got a bit confused about that wrt the highlighting in the report vs the actual supplement content at this point.